Introduce constrain_range; remove old expr_subs

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #95063

This PR introduces a new `constrain_range` function which can be used to constrain the possible values a SymInt/SymFloat can take on. This knowledge can be then used to discharge potential guards (by running the range analysis, and then seeing if the guard must be true given the original range) without adding another guard.

The usage of ranges is very limited right now; ranges are only constrained when the user explicitly instructs the system so. However, we can also infer range constraints based on guards as well; this is left for future work.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

ezyang committed:
Introduce constrain_range; remove old expr_subs

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Introduce constrain_range; remove old expr_subs"


Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Introduce constrain_range; remove old expr_subs"


This PR introduces a new `constrain_range` function which can be used to constrain the possible values a SymInt/SymFloat can take on. This knowledge can be then used to discharge potential guards (by running the range analysis, and then seeing if the guard must be true given the original range) without adding another guard.

The usage of ranges is very limited right now; ranges are only constrained when the user explicitly instructs the system so. However, we can also infer range constraints based on guards as well; this is left for future work.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Introduce constrain_range; remove old expr_subs"


This PR introduces a new `constrain_range` function which can be used to constrain the possible values a SymInt/SymFloat can take on. This knowledge can be then used to discharge potential guards (by running the range analysis, and then seeing if the guard must be true given the original range) without adding another guard.

The usage of ranges is very limited right now; ranges are only constrained when the user explicitly instructs the system so. However, we can also infer range constraints based on guards as well; this is left for future work.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Introduce constrain_range; remove old expr_subs"


This PR introduces a new `constrain_range` function which can be used to constrain the possible values a SymInt/SymFloat can take on. This knowledge can be then used to discharge potential guards (by running the range analysis, and then seeing if the guard must be true given the original range) without adding another guard.

The usage of ranges is very limited right now; ranges are only constrained when the user explicitly instructs the system so. However, we can also infer range constraints based on guards as well; this is left for future work.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

eellison commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -116,6 +118,23 @@ def guard_scalar(a):
+    if not isinstance(a, SymInt):
+        assert min <= a <= max
+        return
+    if isinstance(a.node.expr, sympy.Integer):
```
any reason not to support Floats here ?

ezyang replied:
Same deal, not supporting constraints on floats for now

eellison reviewed (approved):
Nice. We could start producing a range for Symints like `.nonzero()` at some point.

ezyang committed:
Update on "Introduce constrain_range; remove old expr_subs"


This PR introduces a new `constrain_range` function which can be used to constrain the possible values a SymInt/SymFloat can take on. This knowledge can be then used to discharge potential guards (by running the range analysis, and then seeing if the guard must be true given the original range) without adding another guard.

The usage of ranges is very limited right now; ranges are only constrained when the user explicitly instructs the system so. However, we can also infer range constraints based on guards as well; this is left for future work.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented on torch/utils/_sympy/interp.py:
```diff
@@ -66,7 +66,7 @@ def sympy_interp(
     if isinstance(expr, sympy.Integer):
         return analysis.constant(int(expr), torch.int64)
-    elif isinstance(expr, sympy.Float):
+    elif isinstance(expr, sympy.Number):
```
1/3 turns into a Rational which is not a float. There's something kind of fishy going on here... I am still feeling like I want to pass the sympy expressions directly to analysis.constant without casting it to float first.

ezyang commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -1070,6 +1089,11 @@ def __init__(self, allow_scalar_outputs=True):
+        # are conservative: the int MUST fall in the range, but the
+        # range may contain ints which may not actually appear in
+        # practice
+        self.var_to_range: Dict["sympy.Symbol", ValueRanges] = {}
```
It just didn't seem necessary! Also, I view the range analysis on floats as a sort of necessary evil when users do naughty things like sqrt, and for the most part, I only really want to be reasoning about size variables.

ezyang committed:
Update on "Introduce constrain_range; remove old expr_subs"


This PR introduces a new `constrain_range` function which can be used to constrain the possible values a SymInt/SymFloat can take on. This knowledge can be then used to discharge potential guards (by running the range analysis, and then seeing if the guard must be true given the original range) without adding another guard.

The usage of ranges is very limited right now; ranges are only constrained when the user explicitly instructs the system so. However, we can also infer range constraints based on guards as well; this is left for future work.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang committed:
Update on "Introduce constrain_range; remove old expr_subs"


This PR introduces a new `constrain_range` function which can be used to constrain the possible values a SymInt/SymFloat can take on. This knowledge can be then used to discharge potential guards (by running the range analysis, and then seeing if the guard must be true given the original range) without adding another guard.

The usage of ranges is very limited right now; ranges are only constrained when the user explicitly instructs the system so. However, we can also infer range constraints based on guards as well; this is left for future work.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented on torch/utils/_sympy/interp.py:
```diff
@@ -81,8 +81,9 @@ def sympy_interp(
-    if handler in ASSOCIATIVE_OPS:
+    handler_name = handlers()[expr.func]
+    handler = getattr(analysis, handler_name)
+    if handler_name in ASSOCIATIVE_OPS:
```
The old code was just buggy

ezyang committed:
Update on "Introduce constrain_range; remove old expr_subs"


This PR introduces a new `constrain_range` function which can be used to constrain the possible values a SymInt/SymFloat can take on. This knowledge can be then used to discharge potential guards (by running the range analysis, and then seeing if the guard must be true given the original range) without adding another guard.

The usage of ranges is very limited right now; ranges are only constrained when the user explicitly instructs the system so. However, we can also infer range constraints based on guards as well; this is left for future work.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
@pytorchbot merge -f "master lint failure only"

ezyang commented:
@pytorchbot merge -r master

ezyang committed:
Update on "Introduce constrain_range; remove old expr_subs"


This PR introduces a new `constrain_range` function which can be used to constrain the possible values a SymInt/SymFloat can take on. This knowledge can be then used to discharge potential guards (by running the range analysis, and then seeing if the guard must be true given the original range) without adding another guard.

The usage of ranges is very limited right now; ranges are only constrained when the user explicitly instructs the system so. However, we can also infer range constraints based on guards as well; this is left for future work.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

[ghstack-poisoned]

ezyang commented:
oops this broke convit_base

jeanschmidt commented:
@pytorchboot revert -m "Breaking internal builds, more details can be found: https://fburl.com/phabricator/fq5b6k8a" -c nosignal

jeanschmidt commented:
@pytorchbot revert -m "Breaking internal builds, more details can be found: https://fburl.com/phabricator/fq5b6k8a" -c nosignal
